### PR TITLE
feat(commit)!: pass footer token and separator to template

### LINKED ISSUE #96: Commit footers have values but no names
**Describe the bug**

When a conventional commit has footers (such as `Signed-off-by: Someone <email>`), the `footers` array provided to the changelog template contains the value (the part *after* the `:`) but not the name of the footer.  This makes it difficult to implement any kind of special templating based on the name of the footer. For example, if I want to add `Signed-off-by` emails to a changelog entry, but not `Reviewed-by`, it's currently impossible to determine what commit footer that value came from.

**To Reproduce**

In a repository where commits contain multiple footers with different names, create a template with `{{ commit.footers }}` in it somewhere. Note that only the values of the footer are output.

**Expected behavior**
 
The entire footer (including the name) should be exposed to the template. This could either be by making each entry in the `commit.footers` array be the _entire_ footer value, including both the name and the value after the colon, or by changing `commit.footers` from an array to a map of footer names to footer values.

**System**
 - OS Information: 
   ```
   :; uname -a
   Linux noctis 5.17.3 #1-NixOS SMP PREEMPT Wed Apr 13 17:27:43 UTC 2022 x86_64 GNU/Linux
   ```

 - Project Version:
   ```
   :; git-cliff --version
   git-cliff 0.7.0
   ```

**Additional context**

While looking at the code for serializing commits as JSON, I noticed that the `Serialize` impl maps over the list of footers and calls `f.value()`:
https://github.com/orhun/git-cliff/blob/7d0786ca55423950f0779de4e6a907fc25ae203a/git-cliff-core/src/commit.rs#L221-L226

Although I'm not familiar with the `git_conventional` crate's API, my assumption is that this is discarding the footer's name. :)

I'm happy to open a PR changing this behavior, which I imagine would be fairly straightforward. The one design decision is what we should output instead: is it preferable to emit the footers as a map of footer name to footer value (which could break existing templates, but seems nicer), or as an array of the footer's name and value as a single string? Or, is there some other approach that I haven't thought of?

### PR DESCRIPTION
**Pull request by hawkw**
## Description

This branch changes the `Serialize` impl for `Commit` to output conventional
commit footers as structured JSON objects, rather than as just the footer's
value string.

See hawkw/git-cliff@93779b21667fd5f7ec28635e36e428035d69afb5 for a discussion of
the implementation details of this change.

## Motivation and Context

As discussed in issue #96, using footers in changelog templates is currently not
practical in most cases, since there is currently no way to determine *which*
footer a value belongs to, and no way to access the name of the footer.

This PR implements the approach suggested by @orhun in https://github.com/orhun/git-cliff/issues/96#issuecomment-1164883227.
Additionally, I added a section to the README documentation discussing the
structure of the footers passed to templates.

Fixes #96

## How Has This Been Tested?

I added a new test in `commit.rs` that conventional commits with footers have a
`footers` iterator containing the correct footers. In addition, I did some
manual testing to test the use of footers in templates --- see the next section
for details.

## Screenshots / Output (if appropriate):

To manually verify that everthing works end-to-end, I temporarily modified
`examples/detailed.toml` to output footers in the template:

```patch
diff --git a/examples/detailed.toml b/examples/detailed.toml
index b6b57fd..6fc2252 100644
--- a/examples/detailed.toml
+++ b/examples/detailed.toml
@@ -24,6 +24,9 @@ body = """
     ### {{ group | upper_first }}
     {% for commit in commits %}
         - {{ commit.message | upper_first }} ([{{ commit.id | truncate(length=7, end="") }}]({{ commit.id }}))\
+          {% for footer in commit.footers -%}
+            , {{ footer.token }}{{ footer.separator }} {{ footer.value }}\
+          {% endfor %}
     {% endfor %}
 {% endfor %}\n
 """
```

Generating a test changelog with that template now outputs footers, as expected:

```markdown
# Changelog

All notable changes to this project will be documented in this file.

## [unreleased]

### Documentation

- Clarify that `--tag` argument can be an unexisting tag ([d540f5d](d540f5d8938bc84b01b4fafaa69c3290eb72cd08))

- Discuss footers in README ([a54bca5](a54bca57acc91a886d52bb9d3d87866eb5c18f32)), Signed-off-by: Eliza Weisman <eliza@buoyant.io>


### Features

- Support external commands for commit preprocessors (#86) ([7d0786c](7d0786ca55423950f0779de4e6a907fc25ae203a))

- Support changing commit scope with `commit_parsers` (#94) ([e220768](e22076843b91be3680617db5686e070dedcfef29)), Co-authored-by: Orhun Parmaksız <orhunparmaksiz@gmail.com>

...
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other 

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: git-cliff-core/src/commit.rs [hunk 1]
```diff
@@ -263,6 +298,17 @@ impl Serialize for Commit<'_> {
 	}
 }
 
+impl<'a> From<&'a git_conventional::Footer<'a>> for Footer<'a> {
```
Comment (orhun) [inline git-cliff-core/src/commit.rs:301]: Can we move this implementation right after the struct definition of `Footer`?

File: git-cliff-core/src/commit.rs [hunk 2]
```diff
@@ -208,28 +226,45 @@ impl Commit<'_> {
...
+	}
 }
 
 impl Serialize for Commit<'_> {
 	fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
 	where
 		S: Serializer,
 	{
+		/// A wrapper to serialize commit footers from an iterator using
+		/// `Serializer::collect_seq` without having to
```
Comment (orhun) [inline git-cliff-core/src/commit.rs:247]: This seems like an unfinished comment, I didn't quite understand what it means, can you update it?

File: README.md [hunk 3]
```diff
@@ -684,6 +691,31 @@ following context is generated to use for templating:
...
+by an object with the following fields:
+
+- `"token"`, the name of the footer (preceeding the separator character)
+- `separator`, the footer's separator string (either `: ` or ` #`)
+- `value`, the value following the separator character
+- `breaking`, which is `true` if this is a `BREAKING CHANGE:` footer, and
+  `false` otherwise
+
+[_footers_]: https://www.conventionalcommits.org/en/v1.0.0/#specification
+[trailers]: https://git-scm.com/docs/git-interpret-trailers)
```
Comment (orhun) [inline README.md:717]: This doesn't seem to be rendered correctly due to trailing ")" character. Also, can you use in-place links (`[alt](url)`) instead to make it aligned with the rest of README.md?

**Review by orhun — CHANGES_REQUESTED** [review]
Had a couple of nitpicks here and there but looks great! 💯 

Can you also add test fixtures for this change as well? I might do it after merge if you don't want to include in the PR, no problem.

**Comment by orhun** [comment]
> ```diff
> diff --git a/examples/detailed.toml b/examples/detailed.toml
> index b6b57fd..6fc2252 100644
> --- a/examples/detailed.toml
> +++ b/examples/detailed.toml
> @@ -24,6 +24,9 @@ body = """
>      ### {{ group | upper_first }}
>      {% for commit in commits %}
>          - {{ commit.message | upper_first }} ([{{ commit.id | truncate(length=7, end="") }}]({{ commit.id }}))\
> +          {% for footer in commit.footers -%}
> +            , {{ footer.token }}{{ footer.separator }} {{ footer.value }}\
> +          {% endfor %}
>      {% endfor %}
>  {% endfor %}\n
>  """
> ```

Ah, also include this change in the PR please. It seems like a good way to demonstrate how footers can be used within template.

**Comment by orhun** [comment]
Hey @hawkw, can you take a look at this PR soon? I would like to include it in the next release of `git-cliff` if possible 🐻

**Comment by hawkw** [comment]
Sorry, I didn't have a chance to follow up sooner. I'll go ahead and address the remaining feedback, thanks!

**Comment by hawkw** [comment]
> > ```diff
> > diff --git a/examples/detailed.toml b/examples/detailed.toml
> > index b6b57fd..6fc2252 100644
> > --- a/examples/detailed.toml
> > +++ b/examples/detailed.toml
> > @@ -24,6 +24,9 @@ body = """
> >      ### {{ group | upper_first }}
> >      {% for commit in commits %}
> >          - {{ commit.message | upper_first }} ([{{ commit.id | truncate(length=7, end="") }}]({{ commit.id }}))\
> > +          {% for footer in commit.footers -%}
> > +            , {{ footer.token }}{{ footer.separator }} {{ footer.value }}\
> > +          {% endfor %}
> >      {% endfor %}
> >  {% endfor %}\n
> >  """
> > ```
> 
> Ah, also include this change in the PR please. It seems like a good way to demonstrate how footers can be used within template.

Sure, happy to do that! One quick question, though: is there an example repository you're using to generate the example changelogs in the README? I would want to re-generate the "detailed" example now that the template has changed.

**Comment by hawkw** [comment]
> Can you also add test fixtures for this change as well? I might do it after merge if you don't want to include in the PR, no problem.



> Can you also add test fixtures for this change as well? I might do it after merge if you don't want to include in the PR, no problem.

I added some tests for commits with footers in `commit.rs`, here: https://github.com/orhun/git-cliff/blob/9f78b0cb5498a1db6129167eee162b97978011cb/git-cliff-core/src/commit.rs#L354-L404

Are there other tests you're referring to? I'm happy to add any other tests that are necessary for this change, but I wasn't sure where else you would want to test this.

**Comment by orhun** [comment]
> I added some tests for commits with footers in `commit.rs`, here:

Yup, it looks good 👍🏼 

 > Are there other tests you're referring to? I'm happy to add any other tests that are necessary for this change, but I wasn't sure where else you would want to test this.

There are fixture tests in `.github/fixtures` for testing the functionality of `git-cliff` by running it with a pre-defined commit history. It is not super important, but nice to have. Like I said, I can work on it after merging the PR if that's something that you'd not prefer to add 🙂

**Comment by hawkw** [comment]
> > Are there other tests you're referring to? I'm happy to add any other tests that are necessary for this change, but I wasn't sure where else you would want to test this.
> 
> There are fixture tests in `.github/fixtures` for testing the functionality of `git-cliff` by running it with a pre-defined commit history. It is not super important, but nice to have. Like I said, I can work on it after merging the PR if that's something that you'd not prefer to add slightly_smiling_face

Oh, I see, I missed those. I'll figure out how to add those tests as well.

**Comment by orhun** [comment]
> Oh, I see, I missed those. I'll figure out how to add those tests as well.

I just added fixture tests and made some tweaks to the `detailed.toml` example. Merging, thanks for the PR!

**Review by orhun — APPROVED** [review]

**Pull request merged**